enterprise/cluster/src/main/java/org/neo4j/cluster/logging/AsyncLogging.java

```diff
@@ -0,0 +1,146 @@
+import org.neo4j.kernel.lifecycle.LifecycleAdapter;
+import org.neo4j.logging.Log;
+import org.neo4j.logging.LogProvider;
+import org.neo4j.logging.NullLogProvider;
+import org.neo4j.logging.async.AsyncLogEvent;
+import org.neo4j.logging.async.AsyncLogProvider;
+
+import static org.neo4j.logging.async.AsyncLogEvent.process;
+import static org.neo4j.unsafe.impl.internal.dragons.FeatureToggles.flag;
+
+public class AsyncLogging extends LifecycleAdapter
+        implements Consumer<AsyncLogEvent>, ThreadFactory, AsyncEvents.Monitor
+{
+    private static final boolean ENABLED = flag( "ENABLED", true );
+
+    public static LogProvider provider( LifeSupport life, LogProvider provider )
+    {
+        if ( ENABLED )
+        {
+            if ( provider instanceof NullLogProvider )
+            {
+                return provider;
+            }
+            return new AsyncLogProvider( life.add(
+                    new AsyncLogging( provider.getLog( AsyncLogging.class ) ) ).eventSender(), provider );
+        }
+        else
+        {
+            return provider;
+        }
+    }
+
+    private final Log metaLog;
+    private final AsyncEvents<AsyncLogEvent> events;
+    private long highCount;
+    private ExecutorService executor;
+
+    AsyncLogging( Log metaLog )
+    {
+        this.metaLog = metaLog;
+        this.events = new AsyncEvents<>( this, this );
+    }
+
+    @Override
+    public void accept( AsyncLogEvent event )
+    {
+        process( event );
+    }
+
+    @Override
+    public void start()
+    {
+        highCount = 0;
+        executor = Executors.newSingleThreadExecutor( this );
+        executor.submit( events );
+    }
+
+    @Override
+    public void stop() throws InterruptedException
+    {
+        events.shutdown();
+        executor.shutdown();
+        events.awaitTermination();
+    }
+
+    @Override
+    public Thread newThread( final Runnable target )
+    {
+        return new Thread( new RerunOnFailure( target ), getClass().getSimpleName() );
+    }
+
+    @Override
+    public void eventCount( long count )
+    {
+        if ( metaLog.isDebugEnabled() )
+        {
+            if ( count > highCount )
+            {
+                metaLog.debug( "High mark increasing from %d to %d events", highCount, count );
+                highCount = count;
+            }
+        }
+    }
+
+    AsyncEventSender<AsyncLogEvent> eventSender()
+    {
+        return events;
+    }
+
+    private static class RerunOnFailure implements Runnable
+    {
+        private final Runnable target;
+
+        RerunOnFailure( Runnable target )
+        {
+            this.target = target;
+        }
+
+        @Override
+        public void run()
+        {
+            for (; ; )
+            {
+                try
+                {
+                    target.run();
+                }
+                catch ( Throwable e )
+                {
+                    e.printStackTrace();
```

Comment:
Perhaps we should just retry a failing log message until succeeding, with some sleeping in between potentially. This way we will not lose logging messages and will not spam anything into the console. In case of failure the JVM is probably quite unhealthy anyway so my assumption here would be more log events wouldn't build up anyway. Or it's caused by OOM in case log events cannot be created anyway and nothing else would work.

Although rather we should have no `RerunOnFailure` at all and simply try-catch-Throwable and spin in `process`.